### PR TITLE
[COR-563] Add nProbe autotuning via FAISS

### PR DESCRIPTION
**Pull request by jbajic**
### Scope & Purpose

Introduce the autuning feature from the FAISS library. The autotuning in FAISS works by first selecting a number of query points and then finding the ground truth for them in the existing dataset. FAISS sweeps the vector index using different search configurations and returns the set of optimal points.
Since the base index for all the indexes is IVF, these are currently supported search parameters that can be tuned:
- nprobe, max_codes: on SearchParametersIVF
- ht (polysemous): on IVFPQSearchParameters for PQ indexes
- quantizer_efSearch: forwarded to an HNSW coarse quantizer via the nested, owned quantizer_params

The autotune operation is expensive since this is the price. The sampling sizing can get a lot of documents for the autotuning, e.g., given the `targetRecall` of 0.9, we need 5977 query points. That means for all 5977 we need to do a full scan over the vector index.
After that, we have 12 trials, which is a full sweep with gradually increasing search parameters. Let's say we only have `nProbe,` we have the following run sweeps 2,4,8,16,32,64,128,256,512,1024, 2048, 4096. Each of the sweeps produces the mean reached recall, and until we have reached the `targetRecall,` we proceed with the next trial.

## API Changes:

POST `/_api/index/<collection>/<index-id>/autotune` — run tuning (optional topK, minRecall; validated in inspection, out-of-range → 400) with a body like this:
```json
{
  "targetRecall": 0.9,
  "topK": 64
}
```
response:
```json
{
    "code": 200,
    "error": false,
    "result": [
        {
            "error": false,
            "operatingPointCount": 5,
            "reachedTargetRecall": true,
            "shard": "col",
            "targetRecall": 0.95,
            "topK": 100
        }
    ]
}
```
GET `/_api/index/<collection>/<index-id>/autotune` — read the persisted operating-point tables.
Booth fan outr across the dbserves in cluster. They avoid agency and the return reponse is like this:
```json
{
  "error": false,
  "code": 200,
  "tunedTables": [
    {
      "topK": 64,
      "targetRecall": 0.9,
      "points": [
        { "recall": 0.91, "faissKey": "nprobe=8,ht=20", "timeSeconds": 0.0021 }
      ]
    }
  ]
}
```
response:
```json
{
    "code": 200,
    "error": false,
    "tunedTables": [
        {
            "points": [
                {
                    "recall": 0.08794,
                    "searchParameters": "nprobe=1",
                    "timeSeconds": 0.064220947265625
                },
                {
                    "recall": 0.17206,
                    "searchParameters": "nprobe=2",
                    "timeSeconds": 0.10166796875
                },
                {
                    "recall": 0.33622,
                    "searchParameters": "nprobe=4",
                    "timeSeconds": 0.170135009765625
                },
                {
                    "recall": 0.65179,
                    "searchParameters": "nprobe=8",
                    "timeSeconds": 0.303451904296875
                },
                {
                    "recall": 0.95,
                    "searchParameters": "nprobe=16",
                    "timeSeconds": 0.57894091796875
                }
            ],
            "targetRecall": 0.95,
            "topK": 100
        }
    ]
}
```

## Search behaviour
If the autotune table contains the two entries:
- topk: 100, targetRecall 0.9
- topk: 50, targetRecall 0.8

This is the expected behaviour:
- `SORT APPROX_NEAR_L2(..., {targetRecall: 0.95}) LIMIT 100`: The query will run but will issue a warning that it cannot use the targetRecall, but it will fall back on the nearest highest
- `SORT APPROX_NEAR_L2(..., {targetRecall: 0.9}) LIMIT 60`: The query will run but will issue a warning that it cannot use the fact that it was not optimized for such topK
- `SORT APPROX_NEAR_L2(..., {targetRecall: 0.9}) LIMIT 110`: The query will fail since it cannot use the autotune table that it has for a higher topK than it has

## Performance testing
The testing was performed on a glove-100-angular dataset on two indices using the following query
```aql
  FOR d IN vectorColl
     OPTIONS {indexHint: "IVFCosineIdx", forceIndexHint: true}
     LET sim = APPROX_NEAR_COSINE(d.vector, @qp, {targetRecall: 0.9})
     SORT sim
     DESC LIMIT 100
     RETURN {k: d.idx, s: sim}
```
Dataset contains 1183514 vectors of dim=100
The autotune parameters were `targetRecall: 0.9` and `topK: 100`

### IVFFlat
  index:        'IVFCosineIdx' (nLists=16384)
  targetRecall: 0.900
  queries:      2000
  | K  | mean recall | min recall | below target | fail %|
|-------|-------------|------------|--------------|-------- |
|    10 |     0.948   |    0.300   |    262/2000  |  13.1% |
|    50 |     0.925   |    0.420   |    522/2000  |  26.1% |
|   100 |     0.909   |    0.450   |    611/2000  |  30.6% |

### IVF{}_HNSW32,SQx8
  index:        'IVFHNSW_SQ8Idx' (nLists=16384)
  targetRecall: 0.900
  queries:      2000
|   K   | mean recall | min recall | below target | fail % |
-------|-------------|------------|--------------|-------- |
|    10 |     0.921   |    0.200   |    346/2000  |  17.3% |
|    50 |     0.907   |    0.380   |    558/2000  |  27.9% |
|   100 |     0.893   |    0.390   |    667/2000  |  33.4% |

Autotuning takes around 1h

### ToDos:
- [x] Separate tune API
- [x] Extract other parameters from tuning
- [x] How to use the results from autotuning, replace defaultNprobe?
- [ ] FILTER is an issue here since the results from tuning are useless
- [x] Sample a different dataset for tuning, do not use the training dataset

### Type of PR

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [x] **Regression tests**
  - [x] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.12.0: *(Please link PR)*
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/COR-563
- [ ] Design document:

### CODE REVIEW

File: arangod/VectorIndex/AutoTuner.cpp [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+  // Comma-separated `name=value` tokens, e.g. "nprobe=8,ht=20".
+  const auto getPair = [](std::string_view token)
+      -> std::pair<std::string_view, std::string_view> {
+    auto const eq = std::ranges::find(token, '=');
+    if (eq == token.end()) {
+      return {};
+    }
+    return {std::string_view(token.begin(), eq),
+            std::string_view(eq + 1, token.end())};
+  };
```
[the remark this answers is not in the document]
Comment (jbajic) [reply arangod/VectorIndex/AutoTuner.cpp:363]: not true for c++20. This makes sense for the previous standard, but we will not go back in time
